File: server/src/application/entities/message-status.ts
import { User } from './user';
import { Message } from './message';

import { Replace } from '@helpers/Replace';
import { Conversation } from './conversation';

interface MessageStatusProps {
  userId: User['_id'];
  messageId: Message['_id'];
  conversationId: Conversation['_id'];
  readAt: Date | null;
  receivedAt: Date | null;
}

export class MessageStatus {
  private props: MessageStatusProps;

  private constructor(props: MessageStatusProps) {
    this.props = props;
  }

  public get userId() {
    return this.props.userId;
  }

  public get messageId() {
    return this.props.messageId;
  }

  public get conversationId() {
    return this.props.conversationId;
  }

  public get readAt() {
    return this.props.readAt;
  }

  public get receivedAt() {
    return this.props.receivedAt;
  }

  public read() {
    this.props.readAt = new Date();
  }

  public receive() {
    this.props.receivedAt = new Date();
  }

  static create(props: Replace<MessageStatusProps, { readAt?: Date; receivedAt?: Date }>) {
    return new MessageStatus({
      ...props,
      readAt: props.readAt || null,
      receivedAt: props.receivedAt || null,
    });
  }
}

File: server/src/ProxyMity.Domain/Entities/MessageStatus.cs
namespace ProxyMity.Domain.Entities;

[Table("message_status")]
public class MessageStatus
{
    public required Ulid UserId { get; set; }
    public required Ulid MessageId { get; set; }
    public required Ulid ConversationId { get; set; }
    public DateTime? ReadAt { get; set; }
    public DateTime? ReceivedAt { get; set; }

    #region Foreign Keys

    [ForeignKey(nameof(UserId))]
    public User User { get; set; }

    [ForeignKey(nameof(MessageId))]
    public Message Message { get; set; }

    [ForeignKey(nameof(ConversationId))]
    public Conversation Conversation { get; set; }

    #endregion

    public void Read() => ReadAt = DateTime.UtcNow;
    public void Receive() => ReceivedAt = DateTime.UtcNow;

    public static MessageStatus Create(Ulid userId, Ulid messageId, Ulid conversation)
    {
        return new MessageStatus()
        {
            UserId = userId,
            MessageId = messageId,
            ConversationId = conversation
        };
    }
}
